src/Microsoft.Health.Dicom.SqlServer/Features/Schema/Migrations/4.sql

```diff
@@ -1587,6 +1691,102 @@ AS
 GO
 
+/***************************************************************************************/
+-- STORED PROCEDURE
+--     AssignReindexingOperation
+--
+-- DESCRIPTION
+--    Assigns the given operation ID to the set of extended query tags, if possible.
+--
+-- PARAMETERS
+--     @extendedQueryTagKeys
+--         * The list of extended query tag keys
+--     @operationId
+--         * The ID for the re-indexing operation
+--     @returnIfCompleted
+--         * Indicates whether completed tags should also be returned
+--
+-- RETURN VALUE
+--     The subset of keys whose operation was successfully assigned.
+/***************************************************************************************/
+CREATE OR ALTER PROCEDURE dbo.AssignReindexingOperation (
+    @extendedQueryTagKeys dbo.ExtendedQueryTagKeyTableType_1 READONLY,
+    @operationId VARCHAR(32),
+    @returnIfCompleted BIT = 0
+)
+AS
+    SET NOCOUNT     ON
+    SET XACT_ABORT  ON
+
+    BEGIN TRANSACTION
+
+        MERGE INTO dbo.ExtendedQueryTagOperation AS XQTO
+        USING
+        (
+            SELECT input.TagKey
+            FROM @extendedQueryTagKeys AS input
+            INNER JOIN dbo.ExtendedQueryTag AS XQT WITH(HOLDLOCK) ON input.TagKey = XQT.TagKey
+            WHERE TagStatus = 0
+        ) AS tags
+        ON XQTO.TagKey = tags.TagKey
+        WHEN NOT MATCHED THEN
+            INSERT (TagKey, OperationId)
+            VALUES (tags.TagKey, @operationId);
+
+        SELECT XQT.*
+        FROM @extendedQueryTagKeys AS input
+        INNER JOIN dbo.ExtendedQueryTag AS XQT WITH(HOLDLOCK) ON input.TagKey = XQT.TagKey
+        LEFT OUTER JOIN dbo.ExtendedQueryTagOperation AS XQTO WITH(HOLDLOCK) ON XQT.TagKey = XQTO.TagKey
+        WHERE (@returnIfCompleted = 1 AND TagStatus = 1) OR (OperationId = @operationId AND TagStatus = 0)
+
+    COMMIT TRANSACTION
+GO
+
+/***************************************************************************************/
+-- STORED PROCEDURE
+--     CompleteReindexing
+--
+-- DESCRIPTION
+--    Annotates each of the specified tags as "completed" by updating their tag statuses and
+--    removing their association to the re-indexing operation
+--
+-- PARAMETERS
+--     @extendedQueryTagKeys
+--         * The list of extended query tag keys
+--
+-- RETURN VALUE
+--     The keys for the completed tags
+/***************************************************************************************/
+CREATE OR ALTER PROCEDURE dbo.CompleteReindexing (
+    @extendedQueryTagKeys dbo.ExtendedQueryTagKeyTableType_1 READONLY
+)
+AS
+    SET NOCOUNT     ON
+    SET XACT_ABORT  ON
+
+    BEGIN TRANSACTION
+
+        DECLARE @updatedKeys AS dbo.ExtendedQueryTagKeyTableType_1
+
+        -- Update the TagStatus of all rows to Completed (1)
+        UPDATE XQT
+        SET TagStatus = 1
+        OUTPUT INSERTED.TagKey
+        INTO @updatedKeys
+        FROM dbo.ExtendedQueryTag AS XQT
+        INNER JOIN @extendedQueryTagKeys AS input ON XQT.TagKey = input.TagKey
```

Comment:
do we want to add lock here so that `TagStatus` cannot be changed by other transactions during this one?
Imaging TagStatus is changed or the tag is removed before the second statement (DELETE), would DELETE still be valid ?  Although I cannot image a way this could happen, but I think make entire transaction really atomic would be a good idea.